File: Provenance/Controller/PVLynxControllerViewController.m
//
//  PVLynxControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 12/21/16.
//  Copyright © 2016 James Addyman. All rights reserved.
//

#import "PVLynxControllerViewController.h"
#import <PVMednafen/MednafenGameCore.h>

@interface PVLynxControllerViewController ()

@end

@implementation PVLynxControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:OELynxButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:OELynxButtonB];
        }
    }
    
    [self.leftShoulderButton setTag:OELynxButtonOption1];
    [self.rightShoulderButton setTag:OELynxButtonOption2];
    [self.selectButton setTag:OELynxButtonOption2];
    [self.startButton setTag:OELynxButtonOption1];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;

    [lynxCore didReleaseLynxButton:OELynxButtonUp forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonDown forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonLeft forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [lynxCore didPushLynxButton:OELynxButtonUp forPlayer:0];
            [lynxCore didPushLynxButton:OELynxButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [lynxCore didPushLynxButton:OELynxButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [lynxCore didPushLynxButton:OELynxButtonUp forPlayer:0];
            [lynxCore didPushLynxButton:OELynxButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [lynxCore didPushLynxButton:OELynxButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [lynxCore didPushLynxButton:OELynxButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [lynxCore didPushLynxButton:OELynxButtonDown forPlayer:0];
            [lynxCore didPushLynxButton:OELynxButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [lynxCore didPushLynxButton:OELynxButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [lynxCore didPushLynxButton:OELynxButtonDown forPlayer:0];
            [lynxCore didPushLynxButton:OELynxButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    
    [lynxCore didReleaseLynxButton:OELynxButtonUp forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonDown forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonLeft forPlayer:0];
    [lynxCore didReleaseLynxButton:OELynxButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didPushLynxButton:[button tag] forPlayer:0];
    
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didReleaseLynxButton:[button tag] forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didPushLynxButton:OELynxButtonOption1 forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didReleaseLynxButton:OELynxButtonOption1 forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didPushLynxButton:OELynxButtonOption2 forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *lynxCore = (MednafenGameCore *)self.emulatorCore;
    [lynxCore didReleaseLynxButton:OELynxButtonOption2 forPlayer:player];
}

@end

File: Provenance/Controller/PVLynxControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PVLynxControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 12/21/16.
//  Copyright © 2016 James Addyman. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVLynxButton {
        get {
            return PVLynxButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVLynxControllerViewController: PVControllerViewController<PVLynxSystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "A" {
                button.buttonTag = .a
            } else if title == "B" {
                button.buttonTag = .b
            }
        }

        selectButton?.buttonTag = .option2
        startButton?.buttonTag = .option1
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {

        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.option1, forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.option1, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.option2, forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.option2, forPlayer: player)
    }
}
